File: leetcode/tree/272_Closest_Binary_Search_Tree_Value_II/codes/stack.cpp
#include <vector>
#include <stack>
#include <cmath>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution {
public:
    vector<int> closestKValues(TreeNode* root, double target, int k) {
        vector<int> res;
        stack<TreeNode*> pred, succ;

        auto initPredecessor = [&](TreeNode* node) {
            while (node) {
                if (node -> val <= target) {
                    pred.push(node);
                    node = node -> right;
                } else {
                    node = node -> left;
                }
            }
        };

        auto initSuccessor = [&](TreeNode* node) {
            while (node) {
                if (node -> val > target) {
                    succ.push(node);
                    node = node -> left;
                } else {
                    node = node -> right;
                }
            }
        };

        auto getNextPred = [&]() {
            TreeNode* node = pred.top(); pred.pop();
            int val = node -> val;
            node = node -> left;
            while (node) {
                pred.push(node);
                node = node -> right;
            }
            return val;
        };

        auto getNextSucc = [&]() {
            TreeNode* node = succ.top(); succ.pop();
            int val = node -> val;
            node = node -> right;
            while (node) {
                succ.push(node);
                node = node -> left;
            }
            return val;
        };

        initPredecessor(root);
        initSuccessor(root);

        for (int i = 0; i < k; ++i) {
            if (pred.empty()) {
                res.push_back(getNextSucc());
            } else if (succ.empty()) {
                res.push_back(getNextPred());
            } else if (abs(pred.top() -> val - target) <= abs(succ.top() -> val - target)) {
                res.push_back(getNextPred());
            } else {
                res.push_back(getNextSucc());
            }
        }

        return res;
    }
};

File: leetcode/tree/272_Closest_Binary_Search_Tree_Value_II/codes/stack.py
from typing import Optional, List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def closestKValues(self, root: Optional[TreeNode], target: float, k: int) -> List[int]:
        def initPredecessorStack(node):
            while node:
                if node.val <= target:
                    pred.append(node)
                    node = node.right
                else:
                    node = node.left

        def initSuccessorStack(node):
            while node:
                if node.val > target:
                    succ.append(node)
                    node = node.left
                else:
                    node = node.right

        def getNextPred():
            node = pred.pop()
            val = node.val
            node = node.left
            while node:
                pred.append(node)
                node = node.right
            return val

        def getNextSucc():
            node = succ.pop()
            val = node.val
            node = node.right
            while node:
                succ.append(node)
                node = node.left
            return val

        pred, succ = [], []
        initPredecessorStack(root)
        initSuccessorStack(root)

        res = []
        for _ in range(k):
            if not pred:
                res.append(getNextSucc())
            elif not succ:
                res.append(getNextPred())
            elif abs(pred[-1].val - target) <= abs(succ[-1].val - target):
                res.append(getNextPred())
            else:
                res.append(getNextSucc())
        return res
